File: solutions/elixir/beer-song/1/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(0) do
    "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
  end

  def verse(number) do
    "#{bottles(number)} of beer on the wall, #{bottles(number)} of beer.\nTake #{pronoun(number)} down and pass it around, #{bottles(number - 1)} of beer on the wall.\n"
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0) do
    Enum.map(range, &verse(&1))
    |> Enum.join("\n")
  end

  @spec bottles(Integer.t()) :: String.t()
  def bottles(count) do
    case count do
      0 -> "no more bottles"
      1 -> "1 bottle"
      count -> "#{count} bottles"
    end
  end

  @spec pronoun(Integer.t()) :: String.t()
  def pronoun(1), do: "it"
  def pronoun(_), do: "one"
end

File: solutions/typescript/beer-song/1/beer-song.ts
const bottles = (count: number) => count == 1 ? 'bottle' : 'bottles';
const countWord = (count: number) => count == 0 ? 'no more' : `${count}`;
const action = (count: number) => {
  const takeDown = count == 1 ? 'it' : 'one';
  const start = count == 1 ? 'Take it down' : 'Take one down';
  if (count == 0) {
    return 'Go to the store and buy some more, 99 bottles of beer on the wall.';
  }
  return `${start} and pass it around, ${countWord(count - 1)} ${bottles(count - 1)} of beer on the wall.`;
}
const capitalize = (word: string) => word.charAt(0).toUpperCase() + word.slice(1);
export function verse(index: number): string {
  const first = capitalize(countWord(index));
  return `${first} ${bottles(index)} of beer on the wall, ${countWord(index)} ${bottles(index)} of beer.
${action(index)}
`;
}

export function sing(
  initialBottlesCount?: number,
  takeDownCount: number = 0
): string {
  const start = initialBottlesCount ?? 99;
  const end = takeDownCount ?? 0;
  const verses = [];
  for (let i = start; i >= end; i--) {
    verses.push(verse(i));
  }
  return verses.join('\n');
}
